DRIVERS-2228 Revert to using the `count` command for `estimatedDocumentCount` (#1187)
mongodb/specifications: source/crud/crud.rst

estimatedDocumentCount
~~~~~~~~~~~~~~~~~~~~~~

The estimatedDocumentCount function is implemented using the ``count`` command
with no query filter, skip, limit, or other options that would alter the

> results. Once again, the only supported option is maxTimeMS.

Reviewer: the "once again" in the original text seemed to reference that this had already been mentioned at the start of the section. so maybe this should be rephrased?

Drivers MUST document that, due to a bug in versions 5.0.0-5.0.7 of MongoDB,
estimatedDocumentCount was not included in v1 of the Stable API, so users of the
Stable API are recommended to upgrade their server version to 5.0.8+ or set
``apiStrict: false`` to avoid encountering errors when using

> estimatedDocumentCount.

Reviewer 2: "Bug" is rather strong wording. `count` was intentionally omitted from the Stable API. It was an oversight that this would break estimatedDocumentCounts on views. I'm not sure how best to word this, but am open to suggestions.

Author: I took a second pass at this, let me know what you think. I also added a section to the Q&A about why it was decided that this change doesn't require a SemVer major version.

estimatedDocumentCount
~~~~~~~~~~~~~~~~~~~~~~

The estimatedDocumentCount function is implemented using the ``count`` command
with no query filter, skip, limit, or other options that would alter the
results. The only supported options are listed in the
``EstimatedDocumentCountOptions`` type defined above.

> 
> Drivers MUST document that, due to a bug in versions 5.0.0-5.0.7 of MongoDB,
> estimatedDocumentCount was not included in v1 of the Stable API, so users of the
> Stable API are recommended to upgrade their server version to 5.0.8+ or set
> ``apiStrict: false`` to avoid encountering errors when using
> estimatedDocumentCount.
> 

Reviewer 3: Can we just delete this entire paragraph? I think the final paragraph in this section better describes the server bug/regression and also clearly instructs drivers on how they should document this.

```suggestion
```

This may also address @JamesKovacs' concerns with referring to a "bug" here in the [preceding thread](https://github.com/mongodb/specifications/pull/1187/files#r854624143).

Reviewer 3: Points noted. I missed that the later paragraph is mostly referring to "release notes". In that case I'll retract this suggestion.

Drivers MUST document that estimatedDocumentCount will perform a slower but
correct count of the documents when performed against a view. In future server
versions, this will be updated to provide a fast but estimated count instead,

> matching the behavior of the command on regular collections.

Reviewer 3: Then I'd suggest we do whatever we can to ensure driver's get notified of a future fix in the server, so we can come back and update this otherwise open-ended caveat.

> matching the behavior of the command on regular collections.

Author: So I talked to Chris again and there is no timeline or ticket (besides definitely after 7.0), so he suggested linking to a central documentation point that will get updated as things change. The shell's documentation page has already been updated to [include this info](https://www.mongodb.com/docs/manual/reference/method/db.collection.estimatedDocumentCount/#mechanics), but I hesitate to link to there from driver docs as it can easily confuse users, especially of the Node driver. Instead, I updated the requirement to link to the [manual page for the `count` command](https://www.mongodb.com/docs/manual/reference/command/count/#behavior) itself, and if we agree with this change, I'll file a DOCS ticket to get that page updated with the same info that the shell's page has.

results. The only supported options are listed in the
``EstimatedDocumentCountOptions`` type defined above.

Drivers MUST document that, due to an oversight in versions 5.0.0-5.0.7 of
MongoDB, the ``count`` command, which estimatedDocumentCount uses in its
implementation, was not included in v1 of the Stable API, and so users of the
Stable API with estimatedDocumentCount are recommended to upgrade their server
version to 5.0.8+ or set ``apiStrict: false`` to avoid encountering errors.

Drivers MUST document that the ``count`` server command is used to implement

> estimatedDocumentCount and that users can visit `this link
> <https://www.mongodb.com/docs/manual/reference/command/count/#behavior>`_ for
> more information about the command's behavior.

Reviewer 3: Feel free to disregard, but a suggestion to use a contextual name instead of "this".

```suggestion
estimatedDocumentCount and that users can find more information via
`Count: Behavior <https://www.mongodb.com/docs/manual/reference/command/count/#behavior>`_.
```

  Server versions before 3.4 were inconsistent about reporting errors for unrecognized command options and may simply ignore them, which means a client-side error is the only way to inform users that such options are unsupported. For unacknowledged writes using OP_MSG, a client-side error is necessary because the server has no chance to return a response (even though a 3.6+ server is otherwise capable of reporting errors for unrecognized options). For unacknowledged writes using legacy opcodes (i.e. OP_INSERT, OP_UPDATE, and OP_DELETE), the message body has no field with which to express these options so a client-side error is the only mechanism to inform the user that such options are unsupported. The spec does not explicitly refer to unacknowledged writes using OP_QUERY primarily because a response document is always returned and drivers generally would not consider using OP_QUERY precisely for that reason.

Q: Why does reverting to using ``count`` instead of ``aggregate`` with
``$collStats`` for estimatedDocumentCount not require a major version bump in
the drivers, even though it might break users of the Stable API?

>   SemVer `allows <https://semver.org/#what-if-i-inadvertently-alter-the-public-api-in-a-way-that-is-not-compliant-with-the-version-number-change-ie-the-code-incorrectly-introduces-a-major-breaking-change-in-a-patch-release>`_ for a library to include a breaking change in minor or patch version if the change is required to fix another accidental breaking change introduced in a prior version and is not expected to further break a large number of users. Given that the original switching to ``$collStats`` was a breaking change due to it not working on views, the number of users using estimatedDocumentCount with ``apiStrict: true`` is small, and the server is backporting adding ``count`` to the Stable API, we felt that this change was acceptable to make in minor version releases of the drivers per the aforementioned allowance in the SemVer spec.

Reviewer 3: ```suggestion
  SemVer `allows <https://semver.org/#what-if-i-inadvertently-alter-the-public-api-in-a-way-that-is-not-compliant-with-the-version-number-change-ie-the-code-incorrectly-introduces-a-major-breaking-change-in-a-patch-release>`_ for a library to include a breaking change in a minor or patch version if the change is required to fix another accidental breaking change introduced in a prior version and is not expected to further break a large number of users. Given that the original switch to ``$collStats`` was a breaking change due to it not working on views, the number of users using estimatedDocumentCount with ``apiStrict: true`` is small, and the server is back-porting the addition of ``count`` to the Stable API, we felt that this change was acceptable to make in minor version releases of the drivers per the aforementioned allowance in the SemVer spec.
```

Author: GH didn't let me auto commit this, so I just manually made this change.
